protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+                    Err(TransportError::MultiaddrNotSupported(addr))
+                }
+                Err(TransportError::Other(err)) => Err(TransportError::Other(EitherError::A(err))),
+            },
+            // Address does contain circuit relay protocol. Dial destination via relay.
+            Ok(RelayedMultiaddr {
+                relay_peer_id,
+                relay_addr,
+                dst_peer_id,
+                dst_addr,
+            }) => {
+                // TODO: In the future we might want to support dialing a relay by its address only.
+                let relay_peer_id = relay_peer_id.ok_or(RelayError::MissingRelayPeerId)?;
+                let relay_addr = relay_addr.ok_or(RelayError::MissingRelayAddr)?;
+                let dst_peer_id = dst_peer_id.ok_or(RelayError::MissingDstPeerId)?;
+
+                let mut to_behaviour = self.to_behaviour;
+                Ok(EitherFuture::Second(
+                    async move {
+                        let (tx, rx) = oneshot::channel();
+                        to_behaviour
+                            .send(TransportToBehaviourMsg::DialReq {
+                                request_id: RequestId::new(),
+                                relay_addr,
+                                relay_peer_id,
+                                dst_addr,
+                                dst_peer_id,
+                                send_back: tx,
+                            })
+                            .await?;
+                        let stream = rx.await?.map_err(|()| RelayError::Connect)?;
+                        Ok(stream)
+                    }
+                    .boxed(),
+                ))
+            }
+        }
+    }
+
+    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner_transport.address_translation(server, observed)
+    }
+}
+
+#[derive(Default)]
+struct RelayedMultiaddr {
+    relay_peer_id: Option<PeerId>,
+    relay_addr: Option<Multiaddr>,
+    dst_peer_id: Option<PeerId>,
+    dst_addr: Option<Multiaddr>,
+}
+
+/// Parse a [`Multiaddr`] containing a [`Protocol::P2pCircuit`].
+///
+/// Returns `Ok(Err(provided_addr))` when passed address contains no [`Protocol::P2pCircuit`].
+///
+/// Returns `Err(_)` when address is malformed.
+fn parse_relayed_multiaddr(
+    addr: Multiaddr,
+) -> Result<Result<RelayedMultiaddr, Multiaddr>, RelayError> {
+    if !addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)) {
+        return Ok(Err(addr));
+    }
+
+    let mut relayed_multiaddr = RelayedMultiaddr::default();
+
+    let mut before_circuit = true;
+    for protocol in addr.into_iter() {
+        match protocol {
+            Protocol::P2pCircuit => {
+                if before_circuit {
+                    before_circuit = false;
+                } else {
+                    return Err(RelayError::MultipleCircuitRelayProtocolsUnsupported);
```

Comment:
Will this eventually be supported?

Reply:
A hypothetical use case would be that I can imagine some server R1 being accessible to only peer A or peer B, but not both. And R1 is accessible by another server R2 which has the complement accessibility of R1 by A & B.

But I see now it's a bit trickier and would be better to implement this with n=1 first.